File: sdks/ios/Sources/Likerts/Branching.swift
import Foundation

public struct SurveyProgress: Sendable, Equatable {
  public let current: Int
  public let total: Int
  public let visited: Int
  public init(current: Int, total: Int, visited: Int) { self.current=current;self.total=total;self.visited=visited }
}
private func effectivePages(_ schema:SurveySchema)->[SurveyPage]{schema.pages ?? [SurveyPage(id:"survey",questionIds:schema.questions.map(\.id))]}
public func pageRoute(_ schema:SurveySchema,answers:[String:Answer])->[Int]{let all=effectivePages(schema),ids=Dictionary(uniqueKeysWithValues:all.enumerated().map{($0.element.id,$0.offset)}),visible=visibleAnswers(schema.questions,answers:answers);var available=Set<String>(),route:[Int]=[],current=0;while current<all.count{route.append(current);available.formUnion(all[current].questionIds);let branch=all[current].branches?.first{conditionMatches($0.when,available.contains($0.when.questionId) ? visible[$0.when.questionId]:nil)};current=branch.flatMap{ids[$0.goToPageId]} ?? current+1};return route}
public func routedAnswers(_ schema:SurveySchema,answers:[String:Answer])->[String:Answer]{var result=visibleAnswers(schema.questions,answers:answers),pass=0;let all=effectivePages(schema);while pass<=schema.questions.count{let reached=Set(pageRoute(schema,answers:result).flatMap{all[$0].questionIds});let filtered=visibleAnswers(schema.questions,answers:result.filter{reached.contains($0.key)});if filtered.count==result.count{return filtered};result=filtered;pass+=1};return [:]}
public final class SurveyFlow: @unchecked Sendable {
  public let schema:SurveySchema
  private var route:[Int]
  private var history:[Int]
  private var current:Int
  private var values:[String:Answer]
  public init(schema:SurveySchema,answers:[String:Answer]=[:]){self.schema=schema;values=routedAnswers(schema,answers:answers);route=pageRoute(schema,answers:values);current=route.first ?? 0;history=[current]}
  public var page:SurveyPage{effectivePages(schema)[current]}
  public var answers:[String:Answer]{values}
  public var progress:SurveyProgress{SurveyProgress(current:current+1,total:effectivePages(schema).count,visited:history.count)}
  public func setAnswer(_ id:String,_ value:Answer?){let old=history;if let value{values[id]=value}else{values.removeValue(forKey:id)};values=routedAnswers(schema,answers:values);route=pageRoute(schema,answers:values);if !route.contains(current){var shared=route.first ?? 0;for index in 0..<min(old.count,route.count){if old[index] != route[index]{break};shared=route[index]};current=shared};history=Array(route.prefix((route.firstIndex(of:current) ?? 0)+1))}
  @discardableResult public func next()->Bool{guard let index=route.firstIndex(of:current),index+1<route.count else{return false};current=route[index+1];history=Array(route.prefix(index+2));return true}
  @discardableResult public func back()->Bool{guard history.count>1 else{return false};history.removeLast();current=history.last!;return true}
}

File: control-plane/public/demo/sdk/branching.js
import { conditionMatches, visibleAnswers } from './index.js';
const effectivePages = (schema) => schema.pages ?? [{ id: 'survey', questionIds: schema.questions.map(q => q.id) }];
export function pageRoute(schema, answers) {
    const pages = effectivePages(schema), byId = new Map(pages.map((page, index) => [page.id, index]));
    const visible = visibleAnswers(schema.questions, answers);
    const available = new Set();
    const route = [];
    let current = 0;
    while (current < pages.length) {
        route.push(current);
        pages[current].questionIds.forEach(id => available.add(id));
        const branch = pages[current].branches?.find(item => conditionMatches(item.when, available.has(item.when.questionId) ? visible[item.when.questionId] : undefined));
        current = branch ? byId.get(branch.goToPageId) ?? pages.length : current + 1;
    }
    return route;
}
export function routedAnswers(schema, answers) {
    let result = visibleAnswers(schema.questions, answers);
    const pages = effectivePages(schema);
    for (let pass = 0; pass <= schema.questions.length; pass++) {
        const reached = new Set(pageRoute(schema, result).flatMap(index => pages[index].questionIds));
        const filtered = visibleAnswers(schema.questions, Object.fromEntries(Object.entries(result).filter(([id]) => reached.has(id))));
        if (Object.keys(filtered).length === Object.keys(result).length)
            return filtered;
        result = filtered;
    }
    throw new Error('Survey route did not stabilize');
}
export class SurveyFlow {
    schema;
    route;
    history;
    current;
    values;
    constructor(schema, initial = {}) {
        this.schema = schema;
        this.values = routedAnswers(schema, { ...initial });
        this.route = pageRoute(schema, this.values);
        this.current = this.route[0] ?? 0;
        this.history = [this.current];
    }
    get page() { return effectivePages(this.schema)[this.current]; }
    get answers() { return { ...this.values }; }
    get progress() { return { current: this.current + 1, total: effectivePages(this.schema).length, visited: this.history.length }; }
    setAnswer(id, value) { const old = this.history; if (value === undefined)
        delete this.values[id];
    else
        this.values[id] = value; this.values = routedAnswers(this.schema, this.values); this.route = pageRoute(this.schema, this.values); if (!this.route.includes(this.current)) {
        let shared = this.route[0] ?? 0;
        for (let i = 0; i < Math.min(old.length, this.route.length) && old[i] === this.route[i]; i++)
            shared = this.route[i];
        this.current = shared;
    } this.history = this.route.slice(0, this.route.indexOf(this.current) + 1); }
    next() { const at = this.route.indexOf(this.current); if (at < 0 || at + 1 >= this.route.length)
        return false; this.current = this.route[at + 1]; this.history = this.route.slice(0, at + 2); return true; }
    back() { if (this.history.length < 2)
        return false; this.history.pop(); this.current = this.history[this.history.length - 1]; return true; }
}
